python/dgl/dataloading/pytorch/__init__.py

```diff
@@ -1,10 +1,65 @@
 """DGL PyTorch DataLoaders"""
 import inspect
+import torch as th
 from torch.utils.data import DataLoader
 from ..dataloader import NodeCollator, EdgeCollator, GraphCollator
 from ...distributed import DistGraph
 from ...distributed import DistDataLoader
 
+class _ScalarDataLoaderIter:
+    def __init__(self, dataset, batch_size, collate_fn, drop_last):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self.collate_fn = collate_fn
+        self.index = 0
+        self.drop_last = drop_last
+
+    def __next__(self):
+        num_items = self.dataset.shape[0]
+        if self.index >= num_items:
+            raise StopIteration
+        end_idx = self.index + self.batch_size
+        if end_idx > num_items:
+            if self.drop_last:
+                raise StopIteration
+            end_idx = num_items
+        batch = self.dataset[self.index:end_idx]
+        if self.collate_fn:
+            batch = self.collate_fn(batch)
+        self.index += self.batch_size
+
+        return batch
+
+class _ScalarDataLoader:
+    """Custom DataLoader to return mini-batches as tensors, rather than as
+    lists. When used inside of the NodeDataLoader, this significantly reduces
+    the overhead. For the case of a batch size of 1024, instead of giving a
+    list of 1024 tensors to the collator, a single tensor of 1024 dimensions
+    is passed in.
+    This implementation supports only minimum set of features.
+    """
```

Comment:
I've switched it to be an IterableDataset wrapper that manually does the batching (https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) in 
89d8fce, instead of using a custom dataloader.

This is quite necessary, as it's about a 3x speedup as compared to using the default path:
![dataloader_issue](https://user-images.githubusercontent.com/63612878/114747873-24efec00-9d06-11eb-8d49-dce040dba96f.png)

Reply:
@jermainewang Talking with some of the PyTorch folks, one workaround for this is in 0a04381, which drops the cost of converting to a tensor on the GPU from ~16.5ms to ~2.5ms. However, using the IterableDataset as I currently am, avoids the cost of conversion entirely, and is ~25% faster end-to-end (bs=1024, hidden=256, fanout=30,30).